pkg/scalers/gcp_stackdriver_scaler.go

```diff
@@ -0,0 +1,198 @@
+package scalers
+
+import (
+	"context"
+	"fmt"
+	"strconv"
+
+	"k8s.io/api/autoscaling/v2beta2"
+	"k8s.io/apimachinery/pkg/api/resource"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/metrics/pkg/apis/external_metrics"
+	logf "sigs.k8s.io/controller-runtime/pkg/log"
+
+	kedautil "github.com/kedacore/keda/v2/pkg/util"
+)
+
+const (
+	defaultStackdriverTargetValue = 5
+)
+
+type stackdriverScaler struct {
+	client   *StackDriverClient
+	metadata *stackdriverMetadata
+}
+
+type stackdriverMetadata struct {
+	projectId   string
+	filter      string
+	targetValue int
+	metricName  string
+
+	gcpAuthorization *gcpAuthorizationMetadata
+}
+
+var gcpStackdriverLog = logf.Log.WithName("gcp_stackdriver_scaler")
+
+// NewStackdriverScaler creates a new stackdriverScaler
+func NewStackdriverScaler(config *ScalerConfig) (Scaler, error) {
+	meta, err := parseStackdriverMetadata(config)
+	if err != nil {
+		return nil, fmt.Errorf("error parsing Stackdriver metadata: %s", err)
+	}
+
+	return &stackdriverScaler{
+		metadata: meta,
+	}, nil
+}
+
+func parseStackdriverMetadata(config *ScalerConfig) (*stackdriverMetadata, error) {
+	meta := stackdriverMetadata{}
+	meta.targetValue = defaultStackdriverTargetValue
+
+	if val, ok := config.TriggerMetadata["projectId"]; ok {
+		if val == "" {
+			return nil, fmt.Errorf("no projectId name given")
+		}
+
+		meta.projectId = val
+	} else {
+		return nil, fmt.Errorf("no projectId name given")
+	}
+
+	if val, ok := config.TriggerMetadata["filter"]; ok {
+		if val == "" {
+			return nil, fmt.Errorf("no filter given")
+		}
+
+		meta.filter = val
+	} else {
+		return nil, fmt.Errorf("no filter given")
+	}
+
+	if val, ok := config.TriggerMetadata["metricName"]; ok {
```

Comment:
Generate this please instead.

Reply:
For this particular scaler there is no parameter from which to automatically generate the metric.
The project_id is too generic and can collide with another scaler and 'filter' is obviously not appropriate

Reply:
That's why we are adding the prefix to the metric in `GenerateMetricNameWithIndex()` this should ensure uniqueness.

Reply:
no it will report `sX-stackdriver-project` where X is the scaler index in the ScaledObject. So uniqueness is assured.

Reply:
Even if each micro service is in a different K8s cluster ?

Reply:
I guess it doesn't matter even if it does, as each cluster in independent of each other

Reply:
even multiple ScaledObjects can have the same metricName. Because both metricName and label is being used to get the correct metric.